### ISSUE TITLE
iconv linking broken on macOS 10.13 and above?

### ISSUE THREAD

**Issue by chdiza** [vim/vim#3665]
On macOS 10.13 (High Sierra) and 10.14 (Mojave), building vim succeeds.  But at the end I for some reason see the final linking step twice:

```
gcc   -L/usr/local/lib 	-o vim objects/arabic.o objects/beval.o objects/buffer.o objects/blowfish.o objects/crypt.o objects/crypt_zip.o objects/dict.o objects/diff.o objects/digraph.o objects/edit.o objects/eval.o objects/evalfunc.o objects/ex_cmds.o objects/ex_cmds2.o objects/ex_docmd.o objects/ex_eval.o objects/ex_getln.o objects/farsi.o objects/fileio.o objects/fold.o objects/getchar.o objects/hardcopy.o objects/hashtab.o  objects/if_cscope.o objects/if_xcmdsrv.o objects/list.o objects/mark.o objects/memline.o objects/menu.o objects/misc1.o objects/misc2.o objects/move.o objects/mbyte.o objects/normal.o objects/ops.o objects/option.o objects/os_unix.o objects/pathdef.o objects/popupmnu.o objects/pty.o objects/quickfix.o objects/regexp.o objects/screen.o objects/search.o objects/sha256.o objects/spell.o objects/spellfile.o objects/syntax.o objects/tag.o objects/term.o objects/terminal.o objects/ui.o objects/undo.o objects/userfunc.o objects/version.o objects/window.o  objects/encoding.o objects/keyboard.o objects/mouse.o objects/parser.o objects/pen.o objects/termscreen.o objects/state.o objects/unicode.o objects/vterm.o        objects/os_macosx.o objects/os_mac_conv.o   objects/channel.o objects/xdiffi.o objects/xemit.o objects/xprepare.o objects/xutils.o objects/xhistogram.o objects/xpatience.o   objects/charset.o objects/json.o objects/main.o objects/memfile.o objects/message.o        -lm -lncurses  -liconv -framework AppKit
link.sh: OK, linking works, let's try omitting a few libraries.
link.sh: See auto/link.log for details.
link.sh: Trying to omit the iconv library...
link.sh: Vim DOES need the iconv library.
link.sh: Using unmodified link command
  gcc   -L/usr/local/lib 	-o vim objects/arabic.o objects/beval.o objects/buffer.o objects/blowfish.o objects/crypt.o objects/crypt_zip.o objects/dict.o objects/diff.o objects/digraph.o objects/edit.o objects/eval.o objects/evalfunc.o objects/ex_cmds.o objects/ex_cmds2.o objects/ex_docmd.o objects/ex_eval.o objects/ex_getln.o objects/farsi.o objects/fileio.o objects/fold.o objects/getchar.o objects/hardcopy.o objects/hashtab.o  objects/if_cscope.o objects/if_xcmdsrv.o objects/list.o objects/mark.o objects/memline.o objects/menu.o objects/misc1.o objects/misc2.o objects/move.o objects/mbyte.o objects/normal.o objects/ops.o objects/option.o objects/os_unix.o objects/pathdef.o objects/popupmnu.o objects/pty.o objects/quickfix.o objects/regexp.o objects/screen.o objects/search.o objects/sha256.o objects/spell.o objects/spellfile.o objects/syntax.o objects/tag.o objects/term.o objects/terminal.o objects/ui.o objects/undo.o objects/userfunc.o objects/version.o objects/window.o  objects/encoding.o objects/keyboard.o objects/mouse.o objects/parser.o objects/pen.o objects/termscreen.o objects/state.o objects/unicode.o objects/vterm.o        objects/os_macosx.o objects/os_mac_conv.o   objects/channel.o objects/xdiffi.o objects/xemit.o objects/xprepare.o objects/xutils.o objects/xhistogram.o objects/xpatience.o   objects/charset.o objects/json.o objects/main.o objects/memfile.o objects/message.o        -lm -lncurses  -liconv -framework AppKit
link.sh: Linked OK
```

I could swear it used to only happen once.  But anyway, that's the least of the concerns.  I looked in link.log like it says, and I find something that seems important:

```
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      _readfile in fileio.o
      _buf_write_bytes in fileio.o
      _my_iconv_open in mbyte.o
      _string_convert_ext in mbyte.o
     (maybe you meant: _my_iconv_open)
  "_iconv_close", referenced from:
      _readfile in fileio.o
      _buf_write in fileio.o
      _utf_find_illegal in mbyte.o
      _my_iconv_open in mbyte.o
      _convert_setup_ext in mbyte.o
  "_iconv_open", referenced from:
      _my_iconv_open in mbyte.o
     (maybe you meant: _my_iconv_open)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Does this mean that my Vim has broken character conversion support?  , `vim --version` shows `+iconv`, but maybe it *shouldn't* given the above failure.

FYI my build flags are:
```
  --disable-gui
  --without-x
  --disable-netbeans
  --disable-nls
```

**Comment by nuko8** [comment]
This sort of things often happens when a Mac on which Vim is built happens to have two different `iconv`'s; one is macOS's own and another is installed by the user, and thus getting the command line `-I` and `-L` flags inconsistent one another because the compiler picks header files and dyld files in accordance with the order of those flags on the command line.

Anyway, try
```
$ otool -L /your/path/to/vim/executable
``` 
and make sure if, among several lines of the output, there's a line looking like
```
/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
```
and, if any, then make sure if the Mac has the `iconv` dylib file that was specified by the output.  If all the checks are passed, there should be no problem at all.

Otherwise, you need to specify which `iconv` ~~do~~ you want to use first and then tweak the flags accordingly.

**Comment by chdiza** [comment]
I do have `/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)` in the output of `otool -L`, and that dylib does exist where the path says it does.

But I have never installed a second `libiconv`, either on the High Sierra machine or the Mojave one.  

So I don't see how vim can be getting confused; I wonder what that failure message is about.

**Comment by nuko8** [comment]
> I do have /usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0) in the output of otool -L, and that dylib does exist where the path says it does.

Then the installed Vim should work well with the `libiconv`.

> But I have never installed a second libiconv, either on the High Sierra machine or the Mojave one.

My comment doesn't cover all the cases and is a single case which usually happens.  The failure seems to be caused by another reason or details unknown to me.

> I don't see how vim can be getting confused;

I, either, as explained below.

> I wonder what that failure message is about.

When the linker links object files (= `objects/*.o`) to yield the executable, it has to resolve symbols at some point which aren't defined in each object file, some of which come from `.dylib` or `.a` files.  That particular message is issued when the linker couldn't find those symbols in object files or the library files specified by `-l` flags that are accessible through the compiler's default search paths (Usually, it includes `/usr/lib` by default.) or the paths specified by `-L` flags.

So, if you always use macOS's `libiconv`, you'll never get messages like that...  Hmm, your package manager automatically installed another `libiconv` to resolve library dependency for you when you installed a piece of software using that manager, didn't it?

**Comment by chdiza** [comment]
> Hmm, your package manager automatically installed another libiconv to resolve library dependency for you when you installed a piece of software using that manager, didn't it?

No.

It occurs to me that perhaps `link.log` is describing what happened when the build process *attempted* to get away without performing some additional links; the log shows why this attempt didn't work, and so then the additional links are then made.  In other words:
```
link.sh: OK, linking works, let's try omitting a few libraries.
link.sh: See auto/link.log for details.
link.sh: Trying to omit the iconv library...
link.sh: Vim DOES need the iconv library.
```
...the attempt described in line 3 fails (with it's reason recorded in link.log), and that's how link.sh knows that "Vim DOES need the iconv library" after all.

If this is correct, then the failure is nothing to worry about.  But I also then wonder why the attempt is even made in the first place, given that I passed no configuration flags that imply I want to build without `+iconv`.

**Comment by nuko8** [comment]
>> Hmm, your package manager automatically installed another libiconv to resolve library dependency for you when you installed a piece of software using that manager, didn't it?
>No.
>
> It occurs to me that perhaps link.log is describing what happened when the build process attempted to get away without performing some additional links; the log shows why this attempt didn't work, and so then the additional links are then made. In other words:
>
> link.sh: OK, linking works, let's try omitting a few libraries.
> link.sh: See auto/link.log for details.
> link.sh: Trying to omit the iconv library...
> link.sh: Vim DOES need the iconv library.
> ...the attempt described in line 3 fails (with it's reason recorded in link.log), and that's how link.sh knows that "Vim DOES need the iconv library" after all.

Seems I now come to seeing what you've been talking about...  While I've been talking about a real linker error which prevents Vim from being built, you've been talking about what was recorded in `auto/link.log` when the build system tried to build Vim without `libiconv`, right?

> If this is correct, then the failure is nothing to worry about. 

Your conjecture cannot be correct any more; this issue should be closed since nothing is broken :)

> But I also then wonder why the attempt is even made in the first place, given that I passed no configuration flags that imply I want to build without +iconv.

I think this issue tracker isn't supposed to answer questions like this.  But if this question is raised with view of suggesting an optimization or improvement to the build system , I think giving an answer to it may be worth doing.

Linking is not necessarily obvious.  

For example, consider the case where one of the libraries on which Vim depends is also dependent on `libiconv`.  Loading the same library twice is obviously inefficient in time and space (rather, problematic), and, unless there's a given criterion for selecting one of the duplicated pieces of code for a symbol, the linker doesn't know which to choose in order to make the symbol table.  Accordingly, it's a good thing to remove unnecessary libraries when they turn out to be so.

In addition, linkers often automatically make links which is not explicitly specified with command lines.  For example, even if you don't specify `-lc` explicitly on the command line when you build a C program, you'll see that the resulting executable is still linked against `libSystem.B` (= The standard C library on macOS, known as `libc` on other environments).  

Moreover, when you build a C program which uses `libm` functions on macOS, you usually don't need to give `-lm` to the command line, which is not the case for some of other environments.

Last but not least, since Vim tries to be multi platform (See `h develop.txt`), its build system has to deal with various platforms.  If we optimized it for every platform, that would make the relevant code a few times bloater, filled with branches of similar code in order to settle subtle differences between environments.

In short, it's easy to write down linker options to resolve dependency, but optimizing the resulting links needs another care.

Although the current build system appears to be redundant when people look at how it works with a specific library of a certain environment, that redundancy actually helps it work well for every linker and every environment.

**Comment by chdiza** [comment]
> you've been talking about what was recorded in auto/link.log when the build system tried to build > Vim without libiconv, right?

Yes; this is in my initial post.

> Your conjecture cannot be correct any more;

I can't tell whether you are saying that my conjecture is correct, or whether you are saying that it has now become clear that it's incorrect.

I will wait to close until I get confirmation that my conjecture is correct.

**Comment by nuko8** [comment]
I meant the conjecture is absolutely correct; therefore, it cannot be correct any more :)

**Comment by chdiza** [comment]
I would say "any more correct", as in "there's no way to be more correct than that".  "Not correct any more" means "is no longer correct".

**Comment by nuko8** [comment]
I thought the auxiliary verb "can" made a crucial difference :P
